pkg/agent/informant.go

```diff
@@ -1059,3 +1143,169 @@ func (s *InformantServer) Upscale(ctx context.Context, logger *zap.Logger, to ap
 	logger.Info("Received successful upscale result")
 	return nil
 }
+
+// MonitorHealthCheck is the equivalent of (*InformantServer).HealthCheck for
+// when we're connected to a monitor.
+//
+// This method MUST be called while holding s.requestLock AND NOT while holding
+// s.runner.lock.
+func (s *InformantServer) MonitorHealthCheck(ctx context.Context, logger *zap.Logger) error {
+	// Locking requestLock is not technically necessary, but it allows for better
+	// serialization. For example, if this function exits, we know that no other
+	// dispatcher calls will be made.
+	s.requestLock.Lock()
+	defer s.requestLock.Unlock()
+
+	tx, rx := util.NewSingleSignalPair[*MonitorResult]()
+	err := s.dispatcher.Call(
+		ctx,
+		tx,
+		api.HealthCheck{},
+	)
+	if err != nil {
+		return err
+	}
+
+	exit := func(err error) {
+		s.runner.lock.Lock()
+		defer s.runner.lock.Unlock()
+		s.exit(InformantServerExitStatus{
+			Err:            err,
+			RetryShouldFix: false,
+		})
+	}
+
+	// Wait for result
+	timeout := time.Second * time.Duration(s.runner.global.config.Monitor.ResponseTimeoutSeconds)
+	timer := time.NewTimer(timeout)
+	defer timer.Stop()
+	select {
+	case res := <-rx.Recv():
+		// A nil pointer means a monitor error occured
+		if res != nil {
+			return nil
+		} else {
+			err := errors.New("monitor experienced an internal error")
+			exit(err)
+			return err
+		}
+	case <-timer.C:
+		err := fmt.Errorf("timed out waiting %v for monitor response", timeout)
+		exit(err)
+		return err
+	}
```

Comment:
A thought: it seems like this control flow here is pretty common. Maybe either (a) bundle that into `Dispatcher` (so it handles the timeouts for you), or make your own wrapper around this?

(the only parts that change seem to be the argument to `s.dispatcher.Call` and the logic handling the result)